### ISSUE TITLE
[Seine][pdx201] Can't enable loud speaker during phone call (VOIP or cellular)

### ISSUE THREAD

**Issue by AngryGami** [sonyxperiadev/bug_tracker#636]
Platform: Seine
Device: pdx201
Kernel version: 4.14
Android version: 10
Software binaries version: SW_binaries_for_Xperia_Android_10.0.7.1_r1_v11a_seine.zip

**Previously working on**
Kirin has no problem with this on Android 10 build.

**Description**
When talking over VOIP or regular cellular and trying to enable speaker - sound from earpiece disappears but sound from speaker doesn't appear.

**Symptoms**
Looking through logs I've noticed following repeating messages when I enable speaker:
```
2020-10-13 16:47:24.682 6207-6584/? D/audio_hw_primary: select_devices: changing use case low-latency-playback output device from(1: handset, acdb 7) to (21: voice-speaker, acdb 101)
2020-10-13 16:47:24.682 6207-6584/? D/audio_hw_primary: disable_audio_route: reset and update mixer path: low-latency-playback voice-handset
2020-10-13 16:47:24.726 6207-6584/? D/msm8974_platform: platform_split_snd_device: snd_device(20) num devices(0) new_snd_devices(0)
2020-10-13 16:47:24.726 6207-6584/? I/msm8974_platform: platform_get_custom_mtmx_params: no matching param with id 0 ip_ch 2 op_ch 2 uc_id 1 snd_dev 20
2020-10-13 16:47:24.727 6207-6584/? D/msm8974_platform: platform_split_snd_device: snd_device(20) num devices(0) new_snd_devices(-236244868)
2020-10-13 16:47:24.727 6207-6584/? I/msm8974_platform: platform_check_and_set_codec_backend_cfg:becf: afe: bitwidth 16, samplerate 48000 channels 2, backend_idx 0 usecase = 1 device (voice-speaker)
2020-10-13 16:47:24.727 6207-6584/? D/msm8974_platform: platform_split_snd_device: snd_device(21) num devices(1) new_snd_devices(0)
2020-10-13 16:47:24.727 6207-6584/? I/msm8974_platform: platform_check_and_set_codec_backend_cfg: new_snd_devices[0] is 21
2020-10-13 16:47:24.727 6207-6584/? I/msm8974_platform: platform_check_codec_backend_cfg:becf: afe: bitwidth 16, samplerate 48000 channels 2, backend_idx 0 usecase = 1 device (voice-speaker)
2020-10-13 16:47:24.727 6207-6584/? D/msm8974_platform: platform_check_codec_backend_cfg:napb: (0) - (deep-buffer-playback)id (0) sr 48000 bw (16) ch (2) device voice-handset
2020-10-13 16:47:24.731 6207-6584/? D/msm8974_platform: platform_check_codec_backend_cfg:becf: updated afe: bitwidth 16, samplerate 48000 channels 2,backend_idx 0 usecase = 1 device (voice-speaker)
2020-10-13 16:47:24.731 6207-6584/? D/msm8974_platform: platform_check_codec_backend_cfg:becf: afe: playback on codec device not supporting native playback set default Sample Rate(48k)
2020-10-13 16:47:24.732 6207-6584/? I/msm8974_platform: platform_check_codec_backend_cfg:becf: afe: Codec selected backend: 0 updated bit width: 16 and sample rate: 48000
2020-10-13 16:47:24.732 6207-6584/? D/audio_hw_primary: check_usecases_codec_backend:becf: force routing 0
2020-10-13 16:47:24.732 6207-6584/? D/audio_hw_primary: check_usecases_codec_backend:becf: (72) check_usecases curr device: voice-speaker, usecase device:voice-handset backends match 0
2020-10-13 16:47:24.732 6207-6584/? D/audio_hw_primary: check_usecases_codec_backend:becf: (72) check_usecases curr device: voice-speaker, usecase device:voice-handset backends match 0
2020-10-13 16:47:24.732 6207-6584/? E/msm8974_platform: platform_check_backends_match: Invalid snd_device = 
2020-10-13 16:47:24.732 6207-6584/? D/audio_hw_primary: check_usecases_codec_backend:becf: (72) check_usecases curr device: voice-speaker, usecase device: backends match 0
2020-10-13 16:47:24.732 6207-6584/? D/audio_hw_primary: check_usecases_codec_backend:becf: check_usecases num.of Usecases to switch 0
2020-10-13 16:47:24.734 6207-6584/? D/msm8974_platform: platform_split_snd_device: snd_device(21) num devices(0) new_snd_devices(0)
2020-10-13 16:47:24.737 6207-6584/? D/msm8974_platform: platform_split_snd_device: snd_device(21) num devices(1) new_snd_devices(0)
2020-10-13 16:47:24.742 6207-6584/? D/msm8974_platform: platform_split_snd_device: snd_device(21) num devices(0) new_snd_devices(0)
2020-10-13 16:47:24.742 6207-6584/? I/msm8974_platform: platform_get_custom_mtmx_params: no matching param with id 0 ip_ch 2 op_ch 2 uc_id 1 snd_dev 21
2020-10-13 16:47:24.742 6207-6584/? D/audio_hw_primary: enable_audio_route: apply mixer and update path: low-latency-playback voice-speaker
2020-10-13 16:47:24.742 6207-6584/? D/audio_hw_primary: select_devices: done
2020-10-13 16:47:24.742 6207-6253/? D/msm8974_platform: platform_split_snd_device: snd_device(21) num devices(0) new_snd_devices(0)
2020-10-13 16:47:24.742 6207-6253/? I/msm8974_platform: platform_get_custom_mtmx_params: no matching param with id 0 ip_ch 2 op_ch 2 uc_id 1 snd_dev 21
2020-10-13 16:47:24.875 6207-6583/? D/audio_hw_primary: out_set_parameters: enter: usecase(12: audio-ull-playback) kvpairs: routing=2
2020-10-13 16:47:27.086 6207-6249/? D/audio_hw_primary: in_set_parameters: enter: kvpairs=input_source=7;routing=-2147483520;top=
2020-10-13 16:47:29.162 6207-6314/? E/audio_hw_primary: out_write: error -5, cannot write stream data: I/O error
2020-10-13 16:47:29.162 6207-6314/? D/audio_hw_primary: out_standby: enter: stream (0xf1cb7000) usecase(0: deep-buffer-playback)
2020-10-13 16:47:29.677 6207-6314/? D/audio_hw_primary: disable_audio_route: reset and update mixer path: deep-buffer-playback voice-handset
2020-10-13 16:47:29.677 6207-6314/? D/msm8974_platform: platform_split_snd_device: snd_device(20) num devices(0) new_snd_devices(0)
2020-10-13 16:47:29.677 6207-6314/? I/msm8974_platform: platform_get_custom_mtmx_params: no matching param with id 0 ip_ch 2 op_ch 2 uc_id 0 snd_dev 20
2020-10-13 16:47:29.677 6207-6314/? D/audio_hw_primary: disable_snd_device: snd_device(20: voice-handset)
2020-10-13 16:47:29.678 6207-6314/? D/msm8974_platform: platform_split_snd_device: snd_device(20) num devices(0) new_snd_devices(0)
2020-10-13 16:47:29.678 6207-6314/? D/audio_hw_primary: out_standby: exit
2020-10-13 16:47:29.720 6207-6584/? D/audio_hw_primary: out_set_parameters: enter: usecase(0: deep-buffer-playback) kvpairs: routing=2
2020-10-13 16:47:29.745 6207-6314/? D/audio_hw_primary: start_output_stream: enter: stream(0xf1cb7000)usecase(0: deep-buffer-playback) devices(0x2) is_haptic_usecase(0)
2020-10-13 16:47:29.746 6207-6314/? D/audio_hw_primary: select_devices for use case (deep-buffer-playback)
2020-10-13 16:47:29.746 6207-6314/? D/audio_hw_primary: select_devices: changing use case deep-buffer-playback output device from(1: handset, acdb 7) to (21: voice-speaker, acdb 101)
2020-10-13 16:47:29.746 6207-6314/? I/msm8974_platform: platform_check_and_set_codec_backend_cfg:becf: afe: bitwidth 16, samplerate 48000 channels 2, backend_idx 0 usecase = 0 device (voice-speaker)
2020-10-13 16:47:29.746 6207-6314/? D/msm8974_platform: platform_split_snd_device: snd_device(21) num devices(1) new_snd_devices(0)
2020-10-13 16:47:29.746 6207-6314/? I/msm8974_platform: platform_check_and_set_codec_backend_cfg: new_snd_devices[0] is 21
2020-10-13 16:47:29.746 6207-6314/? I/msm8974_platform: platform_check_codec_backend_cfg:becf: afe: bitwidth 16, samplerate 48000 channels 2, backend_idx 0 usecase = 0 device (voice-speaker)
2020-10-13 16:47:29.746 6207-6314/? D/msm8974_platform: platform_check_codec_backend_cfg:napb: (0) - (low-latency-playback)id (1) sr 48000 bw (16) ch (2) device voice-speaker
2020-10-13 16:47:29.750 6207-6314/? D/msm8974_platform: platform_check_codec_backend_cfg:becf: updated afe: bitwidth 16, samplerate 48000 channels 2,backend_idx 0 usecase = 0 device (voice-speaker)
2020-10-13 16:47:29.750 6207-6314/? D/msm8974_platform: platform_check_codec_backend_cfg:becf: afe: playback on codec device not supporting native playback set default Sample Rate(48k)
2020-10-13 16:47:29.750 6207-6314/? I/msm8974_platform: platform_check_codec_backend_cfg:becf: afe: Codec selected backend: 0 updated bit width: 16 and sample rate: 48000
2020-10-13 16:47:29.750 6207-6314/? D/audio_hw_primary: check_usecases_codec_backend:becf: force routing 0
2020-10-13 16:47:29.750 6207-6314/? D/audio_hw_primary: check_usecases_codec_backend:becf: (72) check_usecases curr device: voice-speaker, usecase device:voice-speaker backends match 1
2020-10-13 16:47:29.750 6207-6314/? E/msm8974_platform: platform_check_backends_match: Invalid snd_device = 
2020-10-13 16:47:29.750 6207-6314/? D/audio_hw_primary: check_usecases_codec_backend:becf: (72) check_usecases curr device: voice-speaker, usecase device: backends match 0
2020-10-13 16:47:29.750 6207-6314/? E/msm8974_platform: platform_check_backends_match: Invalid snd_device = 
2020-10-13 16:47:29.750 6207-6314/? D/audio_hw_primary: check_usecases_codec_backend:becf: (72) check_usecases curr device: voice-speaker, usecase device: backends match 0
2020-10-13 16:47:29.751 6207-6314/? D/audio_hw_primary: check_usecases_codec_backend:becf: check_usecases num.of Usecases to switch 0
2020-10-13 16:47:29.751 6207-6314/? D/msm8974_platform: platform_split_snd_device: snd_device(21) num devices(0) new_snd_devices(0)
2020-10-13 16:47:29.751 6207-6314/? D/msm8974_platform: platform_split_snd_device: snd_device(21) num devices(0) new_snd_devices(0)
2020-10-13 16:47:29.754 6207-6314/? D/msm8974_platform: platform_split_snd_device: snd_device(21) num devices(1) new_snd_devices(0)
2020-10-13 16:47:29.757 6207-6314/? D/msm8974_platform: platform_split_snd_device: snd_device(21) num devices(0) new_snd_devices(0)
2020-10-13 16:47:29.757 6207-6314/? I/msm8974_platform: platform_get_custom_mtmx_params: no matching param with id 0 ip_ch 2 op_ch 2 uc_id 0 snd_dev 21
2020-10-13 16:47:29.757 6207-6314/? D/audio_hw_primary: enable_audio_route: apply mixer and update path: deep-buffer-playback voice-speaker
2020-10-13 16:47:29.758 6207-6314/? D/audio_hw_primary: select_devices: done
2020-10-13 16:47:29.761 6207-6314/? E/audio_hw_primary: pcm_open_prepare_helper: pcm_prepare returned -1
2020-10-13 16:47:29.772 6207-6314/? D/audio_hw_primary: disable_audio_route: reset and update mixer path: deep-buffer-playback voice-speaker
2020-10-13 16:47:29.772 6207-6314/? D/msm8974_platform: platform_split_snd_device: snd_device(21) num devices(0) new_snd_devices(0)
2020-10-13 16:47:29.772 6207-6314/? I/msm8974_platform: platform_get_custom_mtmx_params: no matching param with id 0 ip_ch 2 op_ch 2 uc_id 0 snd_dev 21
2020-10-13 16:47:29.772 6207-6314/? D/msm8974_platform: platform_split_snd_device: snd_device(21) num devices(0) new_snd_devices(-203775649)
2020-10-13 16:47:29.823 6207-6314/? D/audio_hw_primary: out_standby: enter: stream (0xf1cb7000) usecase(0: deep-buffer-playback)
2020-10-13 16:47:29.823 6207-6314/? D/audio_hw_primary: out_standby: exit
2020-10-13 16:47:29.886 6207-6314/? D/audio_hw_primary: start_output_stream: enter: stream(0xf1cb7000)usecase(0: deep-buffer-playback) devices(0x2) is_haptic_usecase(0)
2020-10-13 16:47:29.886 6207-6314/? D/audio_hw_primary: select_devices for use case (deep-buffer-playback)
2020-10-13 16:47:29.886 6207-6314/? I/msm8974_platform: platform_check_and_set_codec_backend_cfg:becf: afe: bitwidth 16, samplerate 48000 channels 2, backend_idx 0 usecase = 0 device (voice-speaker)
2020-10-13 16:47:29.886 6207-6314/? D/msm8974_platform: platform_split_snd_device: snd_device(21) num devices(1) new_snd_devices(0)
2020-10-13 16:47:29.886 6207-6314/? I/msm8974_platform: platform_check_and_set_codec_backend_cfg: new_snd_devices[0] is 21
2020-10-13 16:47:29.887 6207-6314/? I/msm8974_platform: platform_check_codec_backend_cfg:becf: afe: bitwidth 16, samplerate 48000 channels 2, backend_idx 0 usecase = 0 device (voice-speaker)
2020-10-13 16:47:29.887 6207-6314/? D/msm8974_platform: platform_check_codec_backend_cfg:napb: (0) - (low-latency-playback)id (1) sr 48000 bw (16) ch (2) device voice-speaker
2020-10-13 16:47:29.890 6207-6314/? D/msm8974_platform: platform_check_codec_backend_cfg:becf: updated afe: bitwidth 16, samplerate 48000 channels 2,backend_idx 0 usecase = 0 device (voice-speaker)
2020-10-13 16:47:29.890 6207-6314/? D/msm8974_platform: platform_check_codec_backend_cfg:becf: afe: playback on codec device not supporting native playback set default Sample Rate(48k)
2020-10-13 16:47:29.890 6207-6314/? I/msm8974_platform: platform_check_codec_backend_cfg:becf: afe: Codec selected backend: 0 updated bit width: 16 and sample rate: 48000
2020-10-13 16:47:29.890 6207-6314/? D/audio_hw_primary: check_usecases_codec_backend:becf: force routing 0
2020-10-13 16:47:29.890 6207-6314/? D/audio_hw_primary: check_usecases_codec_backend:becf: (72) check_usecases curr device: voice-speaker, usecase device:voice-speaker backends match 1
2020-10-13 16:47:29.891 6207-6314/? E/msm8974_platform: platform_check_backends_match: Invalid snd_device = 
2020-10-13 16:47:29.891 6207-6314/? D/audio_hw_primary: check_usecases_codec_backend:becf: (72) check_usecases curr device: voice-speaker, usecase device: backends match 0
2020-10-13 16:47:29.891 6207-6314/? E/msm8974_platform: platform_check_backends_match: Invalid snd_device = 
2020-10-13 16:47:29.891 6207-6314/? D/audio_hw_primary: check_usecases_codec_backend:becf: (72) check_usecases curr device: voice-speaker, usecase device: backends match 0
2020-10-13 16:47:29.891 6207-6314/? D/audio_hw_primary: check_usecases_codec_backend:becf: check_usecases num.of Usecases to switch 0
2020-10-13 16:47:29.891 6207-6314/? D/msm8974_platform: platform_split_snd_device: snd_device(21) num devices(0) new_snd_devices(0)
2020-10-13 16:47:29.891 6207-6314/? D/msm8974_platform: platform_split_snd_device: snd_device(21) num devices(0) new_snd_devices(0)
2020-10-13 16:47:29.894 6207-6314/? D/msm8974_platform: platform_split_snd_device: snd_device(21) num devices(1) new_snd_devices(0)
2020-10-13 16:47:29.898 6207-6314/? D/msm8974_platform: platform_split_snd_device: snd_device(21) num devices(0) new_snd_devices(0)
2020-10-13 16:47:29.898 6207-6314/? I/msm8974_platform: platform_get_custom_mtmx_params: no matching param with id 0 ip_ch 2 op_ch 2 uc_id 0 snd_dev 21
2020-10-13 16:47:29.898 6207-6314/? D/audio_hw_primary: enable_audio_route: apply mixer and update path: deep-buffer-playback voice-speaker
2020-10-13 16:47:29.898 6207-6314/? D/audio_hw_primary: select_devices: done
2020-10-13 16:47:29.902 6207-6314/? E/audio_hw_primary: pcm_open_prepare_helper: pcm_prepare returned -1
2020-10-13 16:47:29.915 6207-6314/? D/audio_hw_primary: disable_audio_route: reset and update mixer path: deep-buffer-playback voice-speaker
```
Most notable are these:
```
D/audio_hw_primary: select_devices: changing use case low-latency-playback output device from(1: handset, acdb 7) to (21: voice-speaker, acdb 101)
```
Here it tries to enable voice-speaker device. Then complains:
```
E/msm8974_platform: platform_check_backends_match: Invalid snd_device =
```
Then complains again:
```
 E/audio_hw_primary: pcm_open_prepare_helper: pcm_prepare returned -1
 D/audio_hw_primary: disable_audio_route: reset and update mixer path: deep-buffer-playback voice-speaker
```
and from this moment this pattern of messages repeats until either I switch back to earpiece or end the call.
This most likely some lacking configuration if `mixer_paths.xml`, though I have no idea which.

**Comment by AngryGami** [comment]
>  ... is more of a hack than anything
Does this mean error is elsewhere and you don't know actually where?

**Comment by MarijnS95** [comment]
> Does this mean error is elsewhere and you don't know actually where?

This means that I'm waiting for confirmation from stock: If the path _name_ is to be believed they might use stereo on stock yet that wouldn't work because `RX_CDC_DMA_RX_0` (the backend for earpiece) is never enabled and the thing really is mono.

It'd be fair on SODP though, because we converted the phone to stereo audio :grin:

**Comment by AngryGami** [comment]
Kirin does have following in `mixer_paths.xml`:
```
    <path name="deep-buffer-playback voice-speaker">
        <path name="deep-buffer-playback" />
    </path>
```

**Comment by AngryGami** [comment]
I'm looking at these messages in the log and trying to match them with mixer_paths.xml file:
```
D/audio_hw_primary: enable_audio_route: apply mixer and update path: low-latency-playback voice-speaker
D/audio_hw_primary: select_devices: done
```
and in `mixer_paths.xml` doesn't have this path, but instead have this:
```
    <path name="low-latency-playback voice-speaker-stereo">
        <path name="low-latency-playback" />
    </path>
...
    <path name="low-latency-playback">
        <ctl name="PRI_MI2S_RX Audio Mixer MultiMedia5" value="1" />
        <ctl name="RX_CDC_DMA_RX_0 Audio Mixer MultiMedia5" value="1" />
    </path>
...
```
Does this mean it tries to enable path that is not defined?

**Comment by AngryGami** [comment]
> That's... Correct, yes?

Well... it kinda strange that log from pdx201 contains references to paths that it doesn't have and doesn't work and at same time kirin, that does work, have these paths.

> With the suggested change you should see `low-latency-playback voice-speaker-stereo`

I don't understand why I should see that there... As far as I understand this message it says that system is trying to set path, and this path will determine how sound stream will be routed (i.e. from which source device to which sink device). You suggest that I need to change name of device from `voice-speaker` to `voice-speaker-stereo` - how then system will know that it needs to use different path? Why I can't just change name of the path then? like this:
```
    <path name="low-latency-playback voice-speaker">
        <path name="low-latency-playback" />
    </path>
```
Wouldn't this have same result?

**Comment by MarijnS95** [comment]
@AngryGami I don't really understand what you're trying to ask. You reported a bug, I gave you a proposed fix that, as far as I understand this conversation, has not been tested. Or are you looking for a preview to the elaborate commit message that'll accompany this change once I submit it?

I guess you're looking for the explanation, so here goes (in hopes of you testing the proposed solution in return...):

The HAL has a so-called `backend_tag_table` that is [used](https://github.com/sonyxperiadev/vendor-qcom-opensource-audio-hal-primary-hal/blob/bbcc00db8ccc47eff1782fc58a13b1f610be13d0/hal/msm8974/platform.c#L4150-L4154) to append a backend name to the path that is selected from `mixer_paths`. In case of stock sound devices Sony added their own `SND_DEVICE_(IN|OUT)_SONY_*` and the mapping is hardcoded in the proprietary HAL, but on SODP we rely on the normal audio devices. From stock logs I gathered that a speakerphone call results in `voice-speaker-stereo` appended to the usecase, so they must have this set in their `backend_tag_table`.

In Seine `audio_platform_info` we instead set the `backend_tag_table` entry for `SND_DEVICE_OUT_VOICE_SPEAKER` to `voice-speaker`, which is a path that doesn't exist in `mixer_paths` as you noticed (and stock doesn't use it, so it's no surprise that it's missing).

Now we can do 3 things:

1. `%s/voice-speaker-stereo"/voice-speaker"/` as you're proposing;
2 (more safe) introduce a new 
    ```xml
    <path name="<usecase> voice-speaker">
        <path name="<usecase> voice-speaker-stereo" />
    </path>
    ```

For _every_ usecase.

That's
```command
$ rg voice-speaker-stereo\" -c rootdir/vendor/etc/mixer_paths.xml
12
```
entries :see_no_evil: 

Instead, if we want to stay close to stock and allow us to diff and update `mixer_paths.xml` in the future we want as little divergence as possible, hence the third option:

3. My suggestion to simply set the value that ends up in `backend_tag_table` to `voice-speaker-stereo` like stock does.

**Comment by AngryGami** [comment]
@MarijnS95 Thank you for explanation 👍 This was exactly what I wanted. 
I don't want to just blindly apply changes to code in hope that this will fix my issue - I do want to understand why these changes might work and fix picture in my head if something doesn't seems correct.
This knowledge you have about how mixer_paths works and how different components of this subsystems interact - where did you get this? I googled for weeks trying to find something reasonably comprehensible but there is nothing. Can you suggest something to read about that? (source code is of course ultimate truth, but I want some general understanding of what is happening there before diving into source)

**Comment by MarijnS95** [comment]
> This knowledge you have about how mixer_paths works and how different components of this subsystems interact - where did you get this? I googled for weeks trying to find something reasonably comprehensible but there is nothing. Can you suggest something to read about that? (source code is of course ultimate truth, but I want some general understanding of what is happening there before diving into source)

I guess hard work and determination... Trying to resolve audio issues while reading through `mixer_paths`, `audio_platform_info` and reading through logs gives most of the information; then it's a matter of time before things snap into place and finally make sense. Other weird details such as the existence of `backend_tag_table` - or why some usecases get the device appended while others don't - really only come from reading the code (in this case, if an entry is not hardcoded in `backend_tag_table` nor has a `backend=` in `audio_platform_info` it won't get appended, like what happens with the speakers on `<path name="low-latency-playback">`).

We're pretty lucky that audio is already pretty verbose; this makes it much easier to connect log lines to the code and follow along that way; one can even enable verbose debugging with `LOCAL_CFLAGS += -DDEBUG -DLOG_NDEBUG=0` and get even more information that way.
The HAL is a little messy (especially all the edgecase) but there aren't any abstractions that make the code hard to follow which is nice.

Then it's a matter of collecting enough information on stock (because they use different audio devices), such as logs from every usecase so that we can tell what path should be selected; if we select a different path that's usually the source of issues.

**Comment by AngryGami** [comment]
I can confirm that your fix does work. I can switch to speaker and back during phone call now.

And what about how mixer_paths works? I mean how to read it? What these `<ctl` elements mean and what values that are used there are coming from? Are path names have to be like they are because something expect them to be? Well... from your response it is clear that something does expect this, but is this just a bad design or intended to work this way all the time?  To me it looks strange that some names have to be  a concatenation of strings instead of just being identifier without semantics...

**Comment by MarijnS95** [comment]
> I can confirm that your fix does work. I can switch to speaker and back during phone call now.

Neat, both phone calls and voip? It's very, very likely that it's not true stereo on stock anyway (or rather, impossible)... So I'll go ahead and submit this change later today.

> And what about how mixer_paths works? I mean how to read it? What these `<ctl` elements mean and what values that are used there are coming from? Are path names have to be like they are because something expect them to be? Well... from your response it is clear that something does expect this, but is this just a bad design or intended to work this way all the time? To me it looks strange that some names have to be a concatenation of strings instead of just being identifier without semantics...

`mixer_paths` is really simple: `path` entries are selected by `libaudioroute`, used by the Audio HALs. Nested `<path>`s "select" other top-level `<path>` entries. The set of `ctl`s together have their value applied on the respective ALSA mixer controls (can be seen with `tinymix`).

Why...
- ...the path names are simple concatenations;
- ...they are constantly changed between platforms (there is no "standard');
- ...there are this many weird controls (`<interface> MultiMediaXX`) that have a simple number but a specific meaning and usecase;
- 

... I don't know.

I'm still craving for doing the mainline audio setup in the kernel to understand how it works under the hood, and if there's perhaps room for simplification (ie. without this many meaningless controls).

**Comment by AngryGami** [comment]
> Neat, both phone calls and voip? 

Yep, both.

Thanks for your time and explanations.

**Comment by MarijnS95** [comment]
@AngryGami Is it stereo, too?

**Comment by AngryGami** [comment]
It's hard to tell :) In order to test that I need to produce stereo signal... Do phone calls actually produce stereo signal? I doubt they do and music played nice even before that (though I didn't test if it actually stereo or not.)

**Comment by MarijnS95** [comment]
Afaik the signal should be upmixed to stereo; it's rather easy to tell when covering _the full_ speaker grille with your finger; alternate top and bottom.

**Comment by AngryGami** [comment]
There is no speaker grills on my device :)
![20201014_233341](https://user-images.githubusercontent.com/2470910/96042584-e91abd80-0e6d-11eb-8393-c70711e0d9b1.jpg)
![20201014_233349](https://user-images.githubusercontent.com/2470910/96042586-ea4bea80-0e6d-11eb-8ba9-08812e49fb2b.jpg)
![20201014_233355](https://user-images.githubusercontent.com/2470910/96042592-ec15ae00-0e6d-11eb-9b1b-acb6eb200d78.jpg)

**Comment by MarijnS95** [comment]
:joy: You took a picture of every side where no speakers are :joy: 

Both reside on the front :)

**Comment by AngryGami** [comment]
Both?
![20201015_000447](https://user-images.githubusercontent.com/2470910/96045168-de622780-0e71-11eb-80ae-faede9509f44.jpg)

**Comment by MarijnS95** [comment]
Yes, both. You've heard stereo audio from both speakers, right? Then you know where it comes from, the earpiece on the top and the little indentation on the glass at the bottom.

**Comment by Haxk20** [comment]
One is at the bottom of the glass. On the right side actually. And  second one is clearly visible even from the picture. Its at the top near the front camera.

**Comment by AngryGami** [comment]
Ok... I'll try to check this tomorrow.

**Comment by AngryGami** [comment]
Well... As far as I can tell sound comes from both speakers - top one AND bottom one.

**Comment by Haxk20** [comment]
That is great to hear.
So both speakers are working.
